app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,200 @@
+
+/**
+ * HotspotService is used to add a foreground service for the wifi hotspot.
+ * Created by Adeel Zafar on 07/01/2019.
+ */
+
+public class HotspotService extends Service {
+
+  public static final String ACTION_TURN_ON_AFTER_O = "Turn_on_hotspot_after_oreo";
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  public static final String ACTION_IS_HOTSPOT_ENABLED = "Is_hotspot_enabled";
+  public static final String ACTION_START_SERVER = "start_server";
+  public static final String ACTION_STOP_SERVER = "stop_server";
+
+  private static final int HOTSPOT_NOTIFICATION_ID = 666;
+  private static final String ACTION_STOP = "hotspot_stop";
+  private static final String TAG = "HotspotService";
+  private WifiHotspotManager hotspotManager;
+  private BroadcastReceiver stopReceiver;
+  private NotificationManager notificationManager;
+  private NotificationCompat.Builder builder;
+  private ZimHostCallbacks zimHostCallbacks;
+  private final IBinder serviceBinder = new HotspotBinder();
+  private WebServerHelper webServerHelper;
+
+  @Override public void onCreate() {
+
+    super.onCreate();
+
+    hotspotManager = new WifiHotspotManager(this);
+
+    stopReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        if (intent != null && intent.getAction().equals(ACTION_STOP)) {
+          stopHotspotAndDismissNotification();
+        }
+      }
+    };
+
+    registerReceiver(stopReceiver, new IntentFilter(ACTION_STOP));
+
+    webServerHelper = new WebServerHelper();
+
+    notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+  }
+
+  @Override public int onStartCommand(@NonNull Intent intent, int flags, int startId) {
+    switch (intent.getAction()) {
+
+      case ACTION_IS_HOTSPOT_ENABLED:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          zimHostCallbacks.onHotspotStateReceived(hotspotManager.checkHotspotState());
+        }
+        break;
+
+      case ACTION_TURN_ON_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          hotspotManager.turnOnHotspot(zimHostCallbacks);
+        }
+        break;
+
+      case ACTION_TURN_OFF_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          stopHotspotAndDismissNotification();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        if (!webServerHelper.startServerHelper(
+            intent.getStringArrayListExtra(SELECTED_ZIM_PATHS_KEY))) {
+          zimHostCallbacks.onServerFailedToStart();
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            stopForeground(true);
+            stopSelf();
+            notificationManager.cancel(HOTSPOT_NOTIFICATION_ID);
+          }
+        } else {
+          zimHostCallbacks.onServerStarted(ServerUtils.getSocketAddress());
+          if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
```

Comment:
You do this on less than O? Don't Foreground Services crash on O>= when you don't attach a notification to it inside of a few seconds?

Reply:
On >=O the notification is started right when the hotspot is turned on, within the `ACTION_TURN_ON_AFTER_OREO`. For below Oreo, we're not turning on hotspot programmatically. So for those devices we need to show user the foreground notification after server is started.